src/signal.rs

```diff
@@ -89,7 +90,9 @@ extern "C" fn fish_signal_handler(
         libc::SIGTERM => {
             // Handle sigterm. The only thing we do is restore the front process ID, then die.
             if !observed {
-                safe_restore_term_mode();
+                if let Some(at_exit) = AT_EXIT.get() {
```

Comment:
The problem is this invokes `terminal_protocols_disable_ifn`. There's no documenation or effort to ensure that function, or the functions they call, are async-signal safe. And we certainly should not be invoking functions like `reader_current_data`  from signal handlers.

It's important that we keep the stuff that signal handlers do well-scoped and easy to follow. To only do what's necessary.

I'm working on a larger cleanup of the tty protocol stuff; FYI I'll be re-introducing this change, but in a way that restores